src/v/cloud_topics/level_zero/stm/ctp_stm_state.cc

```diff
@@ -35,6 +35,13 @@ ctp_stm_state::estimate_min_epoch() const noexcept {
 }
 
 void ctp_stm_state::advance_epoch(cluster_epoch epoch, model::offset offset) {
+    vassert(
+      epoch >= _last_seen_epoch,
+      "Observed a non-monotonic epoch sequence {} < {}",
+      epoch,
+      _last_seen_epoch);
+    _last_seen_epoch = epoch;
+
```

Comment:
I think it could be simpler. Before this assignment:
```cpp
 _max_applied_epoch = std::max(
          epoch, _max_applied_epoch.value_or(cluster_epoch{}));
```
you can check that `epoch >= _max_applied_epoch.value_or(cluter_epoch{})`. The assumption there is that the epoch is monotonic so it's always correct to use max. But before `_max_applied_epoch` is assigned you have the opportunity to compare the old value to the new one.

Reply:
The `frontend` only fences the epoch so it updates `_max_seen_epoch`. It represents the request in-flight.
After giving it a second though I think that it's more difficult. The problem is that from the point of view of `ctp_stm` placeholder batches are idempotent. If some part of the log is replayed twice it's not a problem, it's by design. This could happen naturally when the new replica is created and we're sending the snapshot and then replaying the log. The same is true for LRO updates. They are idempotent. This makes it difficult to create a proper monotonicity check.

Reply:
>This makes it difficult to create a proper monotonicity check.

I think what we are aiming for is not that the log is fully monotonically increasing (that would be nice), but that what flows through apply is monotonically increasing. if you can think of ways we might be able to make the check stronger, that would be great. but for the weaker check do you agree that it is sufficient to check the sequence of batches being applied?